sbi/neural_nets/density_estimators/zuko_flow_estimator.py

```diff
@@ -0,0 +1,170 @@
+            theta = self.z_score_theta(theta)
+        if self.z_score_x:
+            x = self.z_score_x(x)
+
+        return theta, x
+
+    def maybe_z_score_theta(
+        self,
+        theta: torch.Tensor,
+    ) -> torch.Tensor:
+        if self.z_score_theta:
+            theta = self.z_score_theta(theta)
+        return theta
+
+    def maybe_z_score_x(
+        self,
+        x: torch.Tensor,
+    ) -> torch.Tensor:
+        if self.z_score_x:
+            x = self.z_score_x(x)
+        return x
+
+    def loss(self, theta: torch.Tensor, x: torch.Tensor, **kwargs) -> torch.Tensor:
+        theta, x = self.maybe_z_score(theta, x)
+
+        # randomly sample the time steps to compare the vector field at different time steps
+        t = torch.rand(theta.shape[:-1], device=theta.device, dtype=theta.dtype)
+        t_ = t[..., None]
+
+        # sample from probability path at time t
+        epsilon = torch.randn_like(theta)
+        theta_prime = (1 - t_) * theta + (t_ + self.eta) * epsilon
+
+        # compute vector field at the sampled time steps
+        vector_field = epsilon - theta
+
+        # todo: is calling forward here the right thing to do?
+        # compute the mean squared error between the vector fields
+        return (self.forward(theta_prime, x, t) - vector_field).pow(2).mean()
+
+    # TODO: rename? apply z_scores
+    def forward(
+        self, theta: torch.Tensor, x: torch.Tensor, t: torch.Tensor
+    ) -> torch.Tensor:
+        # positional encoding of time steps
+        t = self.frequency * t[..., None]
+        t = torch.cat((t.cos(), t.sin()), dim=-1)
+
+        # broadcast to match shapes of theta, x, and t
+        theta, x, t = broadcast(theta, x, t, ignore=1)
+
+        # return the estimated vector field
+        return self.net(torch.cat((theta, x, t), dim=-1))
+
+    def sample(
+        self, sample_shape: torch.Size, condition: torch.Tensor, **kwargs
+    ) -> torch.Tensor:
+        # check for possible batch dimension in the condition
+        if len(condition.shape) > len(self._condition_shape):
+            raise ValueError(
+                "The condition has a batch dimension, which is currently not supported."
+            )
+
+        dist = self.flow(x_o=condition)
+        samples = dist.sample(sample_shape)
+        return samples
+
+    def flow(self, x_o: torch.Tensor) -> Distribution:
+        x_o = self.maybe_z_score_x(x_o)
+
+        transform = zuko.transforms.ComposedTransform(
+            self.z_score_theta,
+            FreeFormJacobianTransform(
+                f=lambda t, theta: self.forward(theta, x_o, t),
+                t0=x_o.new_tensor(0.0),
+                t1=x_o.new_tensor(1.0),
+                phi=(x_o, *self.net.parameters()),
+            ),
+        )
+        return NormalizingFlow(
+            transform=transform,
+            base=DiagNormal(
+                torch.zeros(self.theta_shape), torch.ones(self.theta_shape)
+            ).expand(x_o.shape[:-1]),
```

Comment:
I am not sure here, but in the SBI setting, I think it makes sense to allow for different base distributions as well, as long as they are fast to sample and evaluate. Because the idea is to transform the prior into the posterior no? 
Let's discuss...